lib/models/os.coffee

```diff
@@ -14,23 +14,223 @@ See the License for the specific language governing permissions and
+			.then (isValid) ->
+				if not isValid
+					throw new errors.ResinRequestError('No such device type')
+			.then ->
+				getOsVersions(deviceType)
+			.asCallback(callback)
+
+	###*
+	# @summary Get the max OS version satisfying the given range
+	# @name getMaxSatisfyingVersion
+	# @public
+	# @function
+	# @memberof resin.models.os
+	#
+	# @param {String} deviceType - device type slug
+	# @param {String} versionOrRange - can be one of
+	# * the exact version number,
+	# in which case it is returned if the version is supported,
+	# or `null` is returned otherwise,
+	# * a [semver](https://www.npmjs.com/package/semver)-compatible
+	# range specification, in which case the most recent satisfying version is returned
+	# if it exists, or `null` is returned,
+	# * `'latest'` in which case the most recent version is returned, including pre-releases,
+	# * `'recommended'` in which case the recommended version is returned, i.e. the most
+	# recent version excluding pre-releases, which can be `null` if only pre-release versions
+	# are available,
+	# * `'default'` in which case the recommended version is returned if available,
+	# or `latest` is returned otherwise.
+	# Defaults to `'latest'`.
+	# @fulfil {String|null} - the version number, or `null` if no matching versions are found
+	# @returns {Promise}
+	#
+	# @example
+	# resin.models.os.getSupportedVersions('raspberry-pi').then(function(osVersions) {
+	# 	console.log('Supported OS versions for raspberry-pi', osVersions);
+	# });
+	#
+	# resin.models.os.getSupportedVersions('raspberry-pi', function(error, osVersions) {
+	# 	if (error) throw error;
+	# 	console.log('Supported OS versions for raspberry-pi', osVersions);
+	# });
+	###
+	exports.getMaxSatisfyingVersion = (deviceType, versionOrRange = 'latest', callback) ->
+		callback = findCallback(arguments)
+
+		return isValidDeviceType(deviceType)
+			.then (isValid) ->
+				if not isValid
+					throw new errors.ResinRequestError('No such device type')
+			.then ->
+				exports.getSupportedVersions(deviceType)
+			.then (osVersions) ->
+				getMaxSatisfyingVersion(versionOrRange, osVersions)
+			.asCallback(callback)
+
+	###*
+	# @summary Format the version number
+	# @name formatVersion
+	# @public
+	# @function
+	# @memberof resin.models.os
+	# @description Prepends `v` to the version number and appends `(recommended)`
+	# if the version is the latest recommended OS version.
+	#
+	# @param {String} v - version string, like '1.2.3'
+	# @param {Object} osVersions - the versions info object compatible with
+	# the output of `resin.model.os.getSupportedVersions()`
+	# Unsupported (unpublished) version will result in exception.
+	# @returns {String}
+	#
+	# @example
+	# console.log(resin.models.os.formatVersion('1.2.3', osVersions)) // => 'v1.2.3 (recommended)'
+	###
+	exports.formatVersion = (v, osVersions) ->
+		v = normalizeVersion(v)
+		result = "v#{v}"
+		if v is osVersions.recommended
+			result += ' (recommended)'
+		return result
```

Comment:
I'd rather not. This is a user-facing set of tools to write code that interacts with Resin, we shouldn't turn it into a place for general purpose internal utilities. We could build a separate library for internal utility functions? For now I'd move it downstream into the CLI instead of here, and we can very easily iterate towards nicer solutions later on.